tests/grammars/Charset.g4

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2017-2020 Renata Hodovan, Akos Kiss, Sebastian Kimberk.
+ *
+ * Licensed under the BSD 3-Clause License
+ * <LICENSE.rst or https://opensource.org/licenses/BSD-3-Clause>.
+ * This file may not be copied, modified, or distributed except
+ * according to those terms.
+ */
+
+/*
+ * Tests handling of charsets (especially with escaped characters).
+ */
+
+// TEST-PROCESS: {grammar}.g4 -o {tmpdir}
+// TEST-GENERATE: {grammar}Generator.{grammar}Generator -r start -o {tmpdir}/{grammar}%d.txt
+// TEST-ANTLR: {grammar}.g4 -o {tmpdir}
+// TEST-PARSE: -p {grammar}Parser -l {grammar}Lexer -r start {tmpdir}/{grammar}%d.txt
+
+grammar Charset;
+
+start
+   : SPECIAL_ESCAPE NON_RANGE_DASH UNICODE_ESCAPE RANGE
+   ;
+
+SPECIAL_ESCAPE
+   : [\-\]] [\]] [\-] [a\-] [\]a] [+\-]
+   ;
+
+NON_RANGE_DASH
+	: [-a] [a-] [-]
+	;
```

Comment:
Some also interesting char sets:
- `[--]` (ANTLR warns about '-' being used multiple times, but eventually accepts the char set. '-' twice.)
- `[-\-]` (Again, ANTLR warning, but still OK. Same as above, '-' twice, but one dash escaped, the other not.)
- `[-\--]` (Warning, but still OK. '-' 3x. The middle dash is _not_ a range marker.)
- `[---]` (No warning! This is a totally fine range of '-'..'-', i.e., a dash.)
- `[\--\-]` (Same as above, but this is how it should look like according to ANTLR docs.)

(The "tricky" part in [ANTLR's documentation](https://github.com/antlr/antlr4/blob/master/doc/lexer-rules.md#lexer-rule-elements)  is that it states that "To get ], \, or - you must escape them with \.", but that's actually not true. If the dash is the first or the last character in the char set, it does not need to be escaped. Which leads to the above interestingly valid patterns.)

Reply:
Yes, documentation is outdated. A hyphen `-` is treated as a range operator being placed between two chars. Otherwise, it's treated as an ordinary hyphen. If you want to declare a hyphen between two chars, you should use escaping.